### ISSUE TITLE
webmin expects letsencrypt pem in different location

### ISSUE THREAD

**Issue by j0bb13** [webmin/webmin#436]
We noticed a lot of email from the automatic renewal failing. It seems certbot saves the certificate in a different location than Webmin expects it to do. Previously it worked without a problem.

The log in /var/log/letsencrypt tells us our certificate has been created succesfully:
`2016-10-05 18:12:18,495:DEBUG:certbot.reporter:Reporting to user: Congratulations! Your certificate and chain have been saved at /etc/letsencrypt/live/_domain_-0003/fullchain.pem.`

And we received an email titled Let's Encrypt certificate renewal failed!
`An error occurred requesting a new certificate for _domain_, www._domain_ from Let's Encrypt : Expected certificate file /etc/letsencrypt/keys/cert.pem was not found
`

After 5 emails we hit the maximum amount of certificate requests, so we got this email about 20 times before we stopped the automatic renewal:
`There were too many requests of a given type :: Error creating new cert :: Too many certificates already issued for exact set of domains: _domain_,www._domain_
`

I tested another domain which didn't have a SSL certificate before, certbot seems to work:
```
2016-10-06 10:56:08,138:DEBUG:certbot.storage:Archive directory /etc/letsencrypt/archive/_domain_ and live directory /etc/letsencrypt/live/_domain_ created.
2016-10-06 10:56:08,138:DEBUG:certbot.storage:Writing certificate to /etc/letsencrypt/live/_domain_/cert.pem.
2016-10-06 10:56:08,138:DEBUG:certbot.storage:Writing private key to /etc/letsencrypt/live/_domain_/privkey.pem.
2016-10-06 10:56:08,139:DEBUG:certbot.storage:Writing chain to /etc/letsencrypt/live/_domain_/chain.pem.
2016-10-06 10:56:08,139:DEBUG:certbot.storage:Writing full chain to /etc/letsencrypt/live/_domain_/fullchain.pem.
2016-10-06 10:56:09,750:DEBUG:certbot.storage:Writing new config /etc/letsencrypt/renewal/_domain_.conf.
2016-10-06 10:56:09,752:DEBUG:certbot.reporter:Reporting to user: Congratulations! Your certificate and chain have been saved at /etc/letsencrypt/live/_domain_/fullchain.pem. Your cert will expire on 2017-01-04. To obtain a new or tweaked version of this certificate in the future, simply run letsencrypt-auto again. To non-interactively renew *all* of your certificates, run "letsencrypt-auto renew"

```

But Virtualmin/Webmin tells us he can't find it in the proper location:
```
Requesting a certificate for _domain_, www._domain_ from Let's Encrypt ..
.. request failed : Expected certificate file /etc/letsencrypt/keys/cert.pem was not found
```

I'm not sure if we're using the native webmin support for letsencrypt, I seem to recall we install the one letsencrypt provided and pointed Webmin to it's location. In that case I assume letsencrypt-auto was updated and the behaviour changed. Is there a way to change to the letsencrypt support native to Webmin if that's the case?

**Comment by clawrie** [comment]
I have the same problem... this seems to have occurred since webmin 1.820 was installed on our server.

**Comment by jcameron** [comment]
That's odd, because Webmin gets the path from the `letsencrypt-auto` command's output.

Does the directory `/etc/letsencrypt/keys` exist on your system? If so, does it contain anything?

**Comment by j0bb13** [comment]
`/etc/letsencrypt/keys` exists and has 188 .pem files created on October 5 and 6. 
`0004_key-certbot.pem` through `0191_key-certbot.pem`.

**Comment by deserted** [comment]
Hey guys, issue is actually related to a change in the output of certbot, it's outputting the key location as the first .pem, so the regex to grab the folder is starting from the wrong folder, meaning it then looks in /etc/letsencrypt/keys/ to find cert.pem, fullchain.pem etc, I'll be putting a pull request in tonight with a solution for this

**Comment by deserted** [comment]
@j0bb13 have a look at #437 - have tested and confirmed working on Ubuntu 14.04 and 16.04, but should be general enough to work with any distro supporting the certbot client.

**Comment by jcameron** [comment]
Thanks for the patch! I will include this in the next release of Webmin.

**Comment by deserted** [comment]
@jcameron Bit of a hijack of this, but saw your commit 42157fdfed7c7eaf50d2dee54032680cb785f7c0, using the python based client by default if the native hasn't been run previously.  Personally I would prefer to continue using the native client, but the convenience of being able to run this straight from webmin (and hence virtualmin) is a hard thing to step away from, especially being able to create alongside the initial domain creation in virtualmin.  Are you open to a patch reworking this, to utilise known output points rather than a regex to check success/failure?  

Documentation for certbot contains the following flags that would allow certs to always be placed in a known location, and could then be confirmed to have been created, with a sanity check beforehand to ensure it's not an old cert in place (for renewals).  If these paths could be passed into the function it would also allow for writing the certs directly to the output location in the case of virtualmin SSL certs.

## Path config for certbot
from https://certbot.eff.org/docs/using.html#command-line
```
paths:
  Arguments changing execution paths & servers

  --cert-path CERT_PATH
                        Path to where cert is saved (with auth --csr),
                        installed from or revoked. (default: None)
  --key-path KEY_PATH   Path to private key for cert installation or
                        revocation (if account key is missing) (default: None)
  --fullchain-path FULLCHAIN_PATH
                        Accompanying path to a full certificate chain (cert
                        plus chain). (default: None)
  --chain-path CHAIN_PATH
                        Accompanying path to a certificate chain. (default:
                        None)
```

If you're open to this let me know and I'll knock up a patch

**Comment by jcameron** [comment]
Sure, at patch for this would be great! The main reason I submitted that change to stop using the native client is that I've been burned multiple times by changes in its output format :-(

**Comment by wiserweb** [comment]
This is a really bad bug. I had a client call me this Canadian Thanks Giving morning reporting that their SSL cert was giving errors. Turns out, the renewal date was last night at midnight and it indeed copies the wrong certificate to the virtual site's webroot.

For anyone that is experiencing this issue you can workaround using these steps.

First locate the newly generated certificate. This is typically in `/etc/letsencrypt/live/domain.com-0002`. You will want the files located in the folder with the largest numeric suffix which indicates the newest renewal.

Second, copy the cert.pem and privkey.pem to the affected webroot to manually overwrite the old certificate and key.

```
cp /etc/letsencrypt/live/domain.com-0002/cert.pem /home/domain.com/ssl.cert
cp /etc/letsencrypt/live/domain.com-0002/privkey.pem /home/domain.com/ssl.key
```
Third, ensure the domain in question owns the new certiicates.

```
chown domain:domain /home/domain.com/ssl.cert
chown domain:domain /home/domain.com/ssl.key
```

Fourth, restart your web server Apache/Nginx.

Finally, test to ensure the SSL has taken by visiting the website in question. You can also verify the certificate details in the Manage SSL Certificate panel which should display the proper domain name and a renewal date somewhere 3 months out into the future.

**Comment by wiserweb** [comment]
Perhaps this is obvious, but if one were to manually apply this patch would the patched file in question be overwritten on the next version?

If so, would it be possible to add note to that effect be added somewhere in the documentation (perhaps this page: https://www.virtualmin.com/documentation/developer/contributing) so that it is clear what to expect if one creates patches to their system prior to a new release.

**Comment by deserted** [comment]
Hey wiserweb, that's unfortunately dependant on the OS's package manager, not a decision made by the webmin installation.  I believe for APT (debian, ubuntu, etc) it will, I'm unsure for RPM but it's an easy one to test - grab an older copy of webmin, install on a VM running your OS of choice (rpm -i, dpkg -i etc), modify a file you can keep track of, then upgrade to the latest version.  I've manually applied the patch as I'd prefer to fix the issue for my clients, I'll just review that file next release and ensure it's updated, if not shouldn't be a problem as you can just grab that file from the repo and drop it straight in.

**Comment by jcameron** [comment]
Generally, any patch that gets included in the Webmin Git repository before the next release is built will be included in that release. So this particular fix will by in the 1.821 and later versions.

**Comment by wiserweb** [comment]
@jcameron would the upgrade overwrite the patched file?

**Comment by wiserweb** [comment]
@deserted yes, I manually applied as well. Just wondering if this is something that will take care of itself on upgrade or if it requires manual intervention.

**Comment by deserted** [comment]
@wiserweb That's going to depend on the package manager being used, what OS are you concerned about? 

As an example, lets take Debian-based distros.  APT/DPKG will overwrite any files owned by the package on upgrade, regardless of modifcation, so long as it hasn't been told explicitly not to (via dpkg-divert).  As mentioned this isn't something related to webmin directly, so the question applies as a generalised question against the package management and distro defaults for your OS.  If you're not using a Debian based distro, (I'm not sure on YUM/RPM's behaviour for this, haven't used it for years) for a definitive answer you're best off asking the appropriate maintainers for the packaging system you are using.

**Comment by wiserweb** [comment]
@deserted Using Debian at the moment. 

It wasn't clear to me how that would work. If it's the package manager then that makes sense. I suppose imagine some scripts running like during the installation but your answer makes sense.

**Comment by jcameron** [comment]
@deserted - I tried your suggestion of using the --cert-file flags, but they didn't appear to have any effect :-(

**Comment by deserted** [comment]
@jcameron yeah found the same thing, are you open to additional deps? I've been looking at https://github.com/do-know/Crypt-LE as a better option to directly integrate

**Comment by swelljoe** [comment]
I love that there's a perl version of the client...but, that module has a *lot* of dependencies of its own. So, it'd be not just one extra dependency, but about a dozen, some of which are XS modules (so couldn't be pulled into Webmin core as local deps).

Here's the full list:

    my $base_build = {
                   'Net::SSLeay' => '1.49',
                   'Test::More' => 0,
                   'IO::Socket::SSL' => '1.42',
                   'Time::Piece' => '1.27',
                   'Crypt::OpenSSL::RSA' => '0.28',
                   'MIME::Base64' => '3.11',
                   'Module::Load' => '0.16',
                   'HTTP::Tiny' => '0.042',
                   'Crypt::OpenSSL::PKCS10' => '0.15',
                   'Log::Log4perl' => '1.27',
                   'Crypt::OpenSSL::Bignum' => '0.06',
                   'Crypt::PKCS10' => '1.5',
                   'JSON::MaybeXS' => '1.003005',
                   'Crypt::Format' => '0.02',
                   'Digest::SHA' => 0,
                   'Scalar::Util' => 0,
                   'File::Temp' => 0,
    };

**Comment by revhal** [comment]
I've been installing let's encrypt in webmin 1.820, but why only webmin site are becoming ssl while the domain is not ?

**Comment by futureweb** [comment]
Hi,
on my System (Centos 7) the Symlinks which should point to the Let's Encrypt Files are created broken ... ?

>CT-000-bash-4.2# pwd
>/etc/webmin
> CT-000-bash-4.2# ll let*
lrwxrwxrwx 1 root root 58 Nov 16 12:29 letsencrypt-ca.pem -> ../../archive/servername-0004/chain1.pem
>lrwxrwxrwx 1 root root 57 Nov 16 12:29 letsencrypt-cert.pem -> ../../archive/servername-0004/cert1.pem
>lrwxrwxrwx 1 root root 60 Nov 16 12:29 letsencrypt-key.pem -> ../../archive/servername-0004/privkey1.pem

Correct path would be:
>/etc/letsencrypt/archive/servername-0004/privkey1.pem

Any Ideas on this? Applied https://github.com/webmin/webmin/pull/437 as it refused to work without ...

Thx
Andreas Schnederle-Wagner

**Comment by jmclaren7** [comment]
I seem to be seeing the same thing as @futureweb ... after applying #437 my original issue was solved but now the symlinks created in /etc/webmin are not valid. I've been looking for the code that generates the symlinks but no luck yet.

**Comment by deserted** [comment]
Are you referring to the symljnks in /etc/letsencrpt/live? If so these are
created by the letsencrpt client, not the webmin code

**Comment by jmclaren7** [comment]
Just like @futureweb described, but just to be clear here are the three links I see created after running "Request Certificate" under Webmin Configuration > SSL Encryption > Let's Encrypt:

`/etc/webmin/letsencrypt-ca.pem -> ../../archive/mysub.domain.com-0003/chain1.pem
/etc/webmin/letsencrypt-cert.pem -> ../../archive/mysub.domain.com-0003/cert1.pem
/etc/webmin/letsencrypt-key.pem -> ../../archive/mysub.domain.com-0003/privkey1.pem`

I've removed the links and reran the certificate request from webmin again and the same invalid links are created, If I understand correctly, valid links would look like:

`/etc/webmin/letsencrypt-ca.pem -> /etc/letsencrypt/live/mysub.domain.com-0003/chain.pem
/etc/webmin/letsencrypt-cert.pem -> /etc/letsencrypt/live/mysub.domain.com-0003/cert.pem
/etc/webmin/letsencrypt-key.pem -> /etc/letsencrypt/live/mysub.domain.com-0003/privkey.pem`

I assumed these links were created by webmin seeing as they are created in /etc/webmin and seemingly are meant to point to the files generated by letsencrypt and stored in /etc/letsencrypt/live

Is this letsencrypt causing this? If so what directs LE to create the link under /etc/webmin and what dictates that webmin's ssl settings be populated with the same paths (i.e. /etc/webmin/letsencrypt-ca.pem)

**Comment by deserted** [comment]
This sounds like the symlinks which are created under
/etc/letsencrypt/live/{domain} are just being copied by webmin. Letsencrypt
creates these as relative symlinks and the examples you're showing would be
valid before being copied. Look for where webmin is copying these links, it
should instead symlinks to them

**Comment by jmclaren7** [comment]
Spot on!

I smashed my keyboard until I found what I needed in /usr/share/webmin/webmin/letsencrypt.cgi after changing the '&copy_source_dest' function to 'symlink' everything worked as expected.

`                       &lock_file($miniserv{'keyfile'});
                        #&copy_source_dest($key, $miniserv{'keyfile'});
                        symlink($key, $miniserv{'keyfile'});
                        &unlock_file($miniserv{'keyfile'});

                        $miniserv{'certfile'} = $config_directory.
                                                "/letsencrypt-cert.pem";
                        &lock_file($miniserv{'certfile'});
                        #&copy_source_dest($cert, $miniserv{'certfile'});
                        symlink($cert, $miniserv{'certfile'});
                        &unlock_file($miniserv{'certfile'});

                        if ($chain) {
                                $miniserv{'extracas'} = $config_directory.
                                                        "/letsencrypt-ca.pem";
                                &lock_file($miniserv{'extracas'});
                                #&copy_source_dest($chain, $miniserv{'extracas'});
                                symlink($chain, $miniserv{'extracas'});
                                &unlock_file($miniserv{'extracas'});
                                }
                        else {
                                delete($miniserv{'extracas'});
                                }
`

I also found some similar code in webmin-lib.pl but changing it did nothing, not sure what it's doing there.

So i'm assuming this issue arose because LE changed the storage layout and started using a relative symlink in place of the actual file or otherwise changed the path output... Would you like me to submit a pull request or anything? I'm not too savvy so i'm not sure this is a complete solution.

**Comment by deserted** [comment]
@jcameron is there a reason to copy this into /etc/webmin rather than just linking to the letsencrypt folder?

**Comment by jmclaren7** [comment]
@deserted FYI, I just noticed that this might of been addressed by @jcameron already with 0e38b9e7a0e427f1e045d2b69bb8118ad910c50c and 7208412de2d5bef6f603cb1d95a9c45c83a5c8c5

copy_source_dest now has an optional third parameter to account for the source being a link and telling it to use the links destination for the source instead.

Sorry to dig this up seeing as it looks like it's fixed already, I learned a lot in the process anyway.

**Comment by alpha1125** [comment]
so, https://github.com/webmin/webmin/issues/436#issuecomment-267179481 was correct in it's investigation.

I was looking around, and trying to enable SSL for the webmin GUI.  Problem was invalid symlinks.

Changing them to point to the correct ones works great...

Steps to correct/enable:
1. use letsecrypt module.
2. fix symlinks.  [depends on your domain]:  assuming /etc/webmin/
    letsencrypt-ca.pem -> /etc/letsencrypt/archive/<domain>/chain1.pem
    letsencrypt-cert.pem -> /etc/letsencrypt/live/<domain>/cert.pem
   letsencrypt-key.pem -> /etc/letsencrypt/live/<domain>/privkey.pem
3. enable ssl in webmin -> webmin configuratkon -> ssl encryption -> enable SSL


I am not sure how this will respond after a cert update. I'm hoping it'll go well.

**Comment by mvhconsult** [comment]
Any update on this bug? Is something real included in current code?

**Comment by jcameron** [comment]
Is this still a problem in Webmin 1.890 ?

**Comment by HazMatt69** [comment]
Yes, this still is a problem

**Comment by Globerada** [comment]
Hi! Our production certificate is going to expire tonight! We posted the problem on letsencrypt forums.

Any steps to solve it please!

https://community.letsencrypt.org/t/lets-encrypt-creates-new-config-files/95728/27

**Comment by Globerada** [comment]
Until someone knows how to properly fix this, I have manually edited apache config

In virtualmin / Services / Configure SSL Website / Edit directives

Changed:

SSLCertificateFile /home/api/ssl.cert
SSLCertificateKeyFile /home/api/ssl.key
SSLCACertificateFile /home/api/ssl.ca

to

SSLCertificateFile /etc/letsencrypt/live/api.example.com-0002/cert.pem
SSLCertificateKeyFile /etc/letsencrypt/live/api.example.com-0002/privkey.pem
SSLCertificateChainFile /etc/letsencrypt/live/api.example.com-0002/chain.pem

**Comment by iliaross** [comment]
❗️ If the issue isn't resolved, feel free to reopen it with updated details.
